### PR TITLE
Vm controller fix update and status update

### PR DESCRIPTION
**Pull request by xpivarc**
### What this PR does
The controller was not adjusted to `/status` subresource changes. This means we were trying to use `Update` even if the change was in `Status` and would be ultimately ignored. 
Some of the changes (to the `Status`) to `vm` inside the `sync` were also ignored because we never passed the `vm` into `updateStatus` func. This is now fixed.

On top of these functional fixes, the last 2 commits address a few issues in unit tests and fixes false positive tests.

Fixes #

### Release note

```release-note
NONE
```

/hold based on https://github.com/kubevirt/kubevirt/pull/11322

### CODE REVIEW

**Comment by jean-edouard** [comment]
/test pull-kubevirt-unit-test
At least 2 tests seem flaky on my machine:
```
[FAILED] Unexpected call to *kubecli.MockVirtualMachineInstanceInterface.Create([context.Background 0xc003a9a000]) at /root/go/src/kubevirt.io/kubevirt/pkg/virt-controller/watch/vm.go:1199 because: there are no expected calls of the method "Create" for that receiver
[FAILED] Timed out after 1.000s.
  The function passed to Eventually failed at /root/go/src/kubevirt.io/kubevirt/pkg/virt-controller/watch/vm_test.go:6143 with:
  Expected
      <[]string | len:0, cap:0>: []
  to have length 1
```

File: pkg/virt-controller/watch/vm_test.go [hunk 1]
```diff
@@ -460,13 +469,13 @@ var _ = Describe("VirtualMachine", func() {
 				vmiInterface.EXPECT().AddVolume(context.Background(), vmi.ObjectMeta.Name, vm.Status.VolumeRequests[0].AddVolumeOptions)
 			}
 
-			vmInterface.EXPECT().Update(context.Background(), gomock.Any()).Do(func(ctx context.Context, arg interface{}) {
-				Expect(arg.(*virtv1.VirtualMachine).Spec.Template.Spec.Volumes[0].Name).To(Equal("vol1"))
-			}).Return(vm, nil)
+			vmInterface.EXPECT().Update(context.Background(), gomock.Any()).DoAndReturn(func(ctx context.Context, vm *virtv1.VirtualMachine) (*virtv1.VirtualMachine, error) {
```
Comment (jean-edouard) [inline pkg/virt-controller/watch/vm_test.go:472]:
```suggestion
			vmInterface.EXPECT().Update(context.Background(), gomock.Any()).DoAndReturn(
				func(ctx context.Context, vm *virtv1.VirtualMachine) (*virtv1.VirtualMachine, error) {
```
(more below)

File: pkg/virt-controller/watch/vm_test.go [hunk 2]
```diff
@@ -334,24 +334,33 @@ var _ = Describe("VirtualMachine", func() {
...
-		}
-
 		sanityExecute := func(vm *virtv1.VirtualMachine) {
 			added := vm.DeepCopy()
 			controller.Execute()
 			Expect(equality.Semantic.DeepEqual(vm, added)).To(BeTrue(), "A cached VM was modified")
 		}
 
+		expectOnlyInitializedConditionToBeChanged := func(vm *virtv1.VirtualMachine) {
+			vmInterface.EXPECT().UpdateStatus(context.Background(), gomock.Any()).DoAndReturn(func(ctx context.Context, vmArg *virtv1.VirtualMachine) (interface{}, error) {
```
Comment (jean-edouard) [inline pkg/virt-controller/watch/vm_test.go:349]:
```suggestion
			vmInterface.EXPECT().UpdateStatus(context.Background(), gomock.Any()).DoAndReturn(
				func(ctx context.Context, vmArg *virtv1.VirtualMachine) (interface{}, error) {
```

File: pkg/virt-controller/watch/vm_test.go [hunk 3]
```diff
@@ -5640,39 +5879,26 @@ var _ = Describe("VirtualMachine", func() {
...
-				modifyVirtualMachine(vm)
+				addVirtualMachine(vm)
 
 				By("Executing the controller again expecting the RestartRequired condition to appear")
-				expectControllerRevisionDelete()
-				expectVMUpdate()
+				expectVMStatusUpdate()
 				sanityExecute(vm)
-				syncCaches()
 				Eventually(restartRequired, 10*time.Second).Should(BeTrue(), "restart required")
```
Comment (jean-edouard) [inline pkg/virt-controller/watch/vm_test.go:5902]: That's still required? :(

File: pkg/virt-controller/watch/vm_test.go [hunk 4]
```diff
@@ -5640,39 +5879,26 @@ var _ = Describe("VirtualMachine", func() {
...
-				expectVMICreation()
-				expectVMStatusUpdate()
-				expectControllerRevisionList()
-				expectControllerRevisionCreation()
-				sanityExecute(vm)
-				syncCaches()
-				Consistently(restartRequired, 1*time.Second).Should(BeFalse(), "restart required")
-				markAsReady(vmi)
-				vmiFeeder.Add(vmi)
+				vmiSource.Add(vmi)
```
Comment (jean-edouard) [inline pkg/virt-controller/watch/vm_test.go:5888]: Some tests in vmi_test.go use the suggestion below, and seem to have less/no timing issue. Maybe something to explore?
```suggestion
				Expect(vmiInformer.GetIndexer().Add(vmi)).To(Succeed())
```

File: pkg/virt-controller/watch/vm_test.go [hunk 5]
```diff
@@ -5684,111 +5910,109 @@ var _ = Describe("VirtualMachine", func() {
...
-				syncCaches()
 				Eventually(restartRequired, 10*time.Second).Should(BeTrue(), "restart required")
 			})
 
 			DescribeTable("when changing a live-updatable field", func(fgs []string, strat *virtv1.VMRolloutStrategy, expectCond bool) {
+				// Add necessary stuff to reflect running VM
+				// TODO: This should be done in more places
+				vm.Status.Created = true
+				vm.Status.Ready = true
+				vm.Status.PrintableStatus = "Running"
```
Comment (jean-edouard) [inline pkg/virt-controller/watch/vm_test.go:5969]: Can't we set those in `VirtualMachineFromVMI()` when started == true?

**Review by jean-edouard — COMMENTED** [review]
/test pull-kubevirt-unit-test-arm64
/test pull-kubevirt-unit-test
Looks great, some comments in-line. A bit concerned about the flakes I'm seeing...

**Comment by xpivarc** [comment]
/close 
https://github.com/kubevirt/kubevirt/pull/11836

**Pull request closed without merging**